### PR TITLE
SRW: Add arbitrary source

### PR DESCRIPTION
**Pull request by ahebnl**
add the arbitrary magnetic field to Sirepo.

### CODE REVIEW

File: sirepo/server.py [hunk 1]
```diff
@@ -18,7 +18,6 @@
 from sirepo import runner
 from sirepo import runner_client
 from sirepo import simulation_db
-from sirepo import srdb
```
Comment (mrakitin) [inline sirepo/server.py]: I think this should be reverted.

File: sirepo/template/srw.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-u"""SRW execution template.
```
Comment (mrakitin) [inline sirepo/template/srw.py]: I think it still should be with `u`.

File: sirepo/template/srw.py [hunk 3]
```diff
@@ -541,6 +553,7 @@ def lib_files(data, source_lib):
         res.append(dm['mirrorReport']['heightProfileFile'])
     if _uses_tabulated_zipfile(data):
         if 'tabulatedUndulator' in dm and dm.tabulatedUndulator.magneticFile:
+            print('dm.tabulatedUndulator.magneticFile',dm.tabulatedUndulator.magneticFile)
```
Comment (mrakitin) [inline sirepo/template/srw.py]:
```suggestion
            pkdlog('dm.tabulatedUndulator.magneticFile',dm.tabulatedUndulator.magneticFile)
```

File: sirepo/template/srw.py [hunk 4]
```diff
@@ -848,6 +864,7 @@ def write_parameters(data, run_dir, is_parallel):
         run_dir (py.path): where to write
         is_parallel (bool): run in background?
     """
+    print('write_parameters file to {}'.format(run_dir))
```
Comment (mrakitin) [inline sirepo/template/srw.py]:
```suggestion
    pkdlog('write_parameters file to {}'.format(run_dir))
```

File: sirepo/template/srw.py [hunk 5]
```diff
@@ -1405,7 +1421,8 @@ def _generate_parameters_file(data, plot_reports=False, run_dir=None):
     data['models']['intensityReport']['magneticField'] = magnetic_field
     data['models']['sourceIntensityReport']['magneticField'] = magnetic_field
     data['models']['trajectoryReport']['magneticField'] = magnetic_field
-
+    data['models']['powerDensityReport']['magneticField'] = magnetic_field
+    #pkdlog('magnetic_field={}',magnetic_field)
```
Comment (mrakitin) [inline sirepo/template/srw.py]:
```suggestion
```

File: sirepo/template/srw.py [hunk 6]
```diff
@@ -1498,7 +1525,7 @@ def _generate_srw_main(data, plot_reports):
     run_all = report == _RUN_ALL_MODEL
     content = [
         'v = srwl_bl.srwl_uti_parse_options(varParam, use_sys_argv={})'.format(plot_reports),
-        'source_type, mag = srwl_bl.setup_source(v)',
+        #'source_type, mag = srwl_bl.setup_source(v)',
```
Comment (mrakitin) [inline sirepo/template/srw.py]: Why is it commented out?

File: sirepo/template/srw.py [hunk 7]
```diff
@@ -1532,7 +1559,7 @@ def _generate_srw_main(data, plot_reports):
         if plot_reports:
             content.append("v.ws_pl = 'xy'")
     if plot_reports or not _is_background_report(report):
-        content.append('srwl_bl.SRWLBeamline(_name=v.name, _mag_approx=mag).calc_all(v, op)')
+        content.append('srwl_bl.SRWLBeamline(_name=v.name).calc_all(v, op)')
```
Comment (mrakitin) [inline sirepo/template/srw.py]: Why don't we use the `mag` parameter here?

File: sirepo/template/template_common.py [hunk 8]
```diff
@@ -289,6 +289,7 @@ def render_jinja(sim_type, v, name=PARAMETERS_PYTHON_FILE):
         str: source text
     """
     b = resource_dir(sim_type).join(name)
+    #pkdlog("the template is {}", b)
```
Comment (mrakitin) [inline sirepo/template/template_common.py]:
```suggestion
```

File: sirepo/template/srw.py [hunk 9]
```diff
@@ -1489,6 +1507,15 @@ def _generate_parameters_file(data, plot_reports=False, run_dir=None):
             mmz.z.extract(df, target_dir)
         v.magneticMeasurementsDir = _TABULATED_UNDULATOR_DATA_DIR + '/' + mmz.index_dir
         v.magneticMeasurementsIndexFile = mmz.index_file
+
+    # prepare the field file
+    #raise RuntimeError("{}, {}".format( data['models']['simulation']['sourceType'], run_dir))
+    if data['models']['simulation']['sourceType'] == 'a':
+        field_file = str(simulation_db.simulation_lib_dir('srw').join(v['arbitraryMagField_magneticFile']))
+        pkdlog("rundir= {}, {}", run_dir, field_file)
+        pkdlog("lib path= {}", simulation_db.simulation_lib_dir('srw'))
```
Comment (mrakitin) [inline sirepo/template/srw.py]: Do we need this logs in production or is it only for tests?

File: sirepo/template/srw.py [hunk 10]
```diff
@@ -1489,6 +1507,15 @@ def _generate_parameters_file(data, plot_reports=False, run_dir=None):
             mmz.z.extract(df, target_dir)
         v.magneticMeasurementsDir = _TABULATED_UNDULATOR_DATA_DIR + '/' + mmz.index_dir
         v.magneticMeasurementsIndexFile = mmz.index_file
+
+    # prepare the field file
+    #raise RuntimeError("{}, {}".format( data['models']['simulation']['sourceType'], run_dir))
```
Comment (mrakitin) [inline sirepo/template/srw.py]: I'd recommend to clean up the commented lines above.

File: sirepo/template/srw.py [hunk 11]
```diff
@@ -79,6 +80,7 @@
     'mirror': ['dat', 'txt'],
     'sample': ['tif', 'tiff', 'png', 'bmp', 'gif', 'jpg', 'jpeg'],
     'undulatorTable': ['zip'],
+    'arbitraryField': ['dat', 'txt']
```
Comment (mrakitin) [inline sirepo/template/srw.py]:
```suggestion
    'arbitraryField': ['dat', 'txt'],
```

File: sirepo/server.py [hunk 12]
```diff
@@ -719,10 +725,27 @@ def _as_attachment(resp, content_type, filename):
 
 @pkconfig.parse_none
 def _cfg_db_dir(value):
-    """DEPRECATED"""
-    if value is not None:
-        srdb.server_init_root(value)
-    return srdb.root()
```
Comment (mrakitin) [inline sirepo/server.py]: @moellep, I am not sure about this change. Your guidance will be very helpful here. Thanks.

File: sirepo/package_data/template/srw/parameters.py.jinja [hunk 13]
```diff
@@ -78,6 +78,12 @@ varParam = srwl_bl.srwl_uti_ext_options([
     ['und_mfs', 's', '{{magneticMeasurementsIndexFile}}', 'name of magnetic measurements for different gaps summary file'],
 
 {% endif %}
+{% if simulation_sourceType == 'a' %}
+    ['mag_ior','f',{{arbitraryMagField_interpolationOrder}},'interpolation order to use for tabulated magnetic field (1<= mag_ior <= 3)'],
+    ['mag_zc','f',{{arbitraryMagField_longitudinalPosition}},'center longitudinal position [m]'],
+    ['mag_fn','s','{{arbitraryMagField_magneticFile}}','input 3D magnetic field file name'],
```
Comment (mrakitin) [inline sirepo/package_data/template/srw/parameters.py.jinja]: Please put blank spaces after commas to make this PEP8 compliant.

File: sirepo/server.py (hunk 12 above)
Comment (moellep) [reply sirepo/server.py]: @mrakitin, yes, I think it clobbered the srdb.py change - I'll fix this before merging.

Comment (mrakitin) [reply sirepo/server.py]: Thanks Paul! Feel free to push to An's PR branch.

File: sirepo/template/srw.py (hunk 6 above)
Comment (moellep) [reply sirepo/template/srw.py]: @mrakitin, I think this and the ```mag``` parameter change below have been integrated into SRW

**Comment by moellep** [comment]
Thanks @ahebnl, I've merged the changes. I did change a few of the pkdlog() statements into pkdc(). You can enable the pkdc() output by setting the trace environment variable before running the http server:
```
export PYKERN_PKDEBUG_CONTROL=srw
sirepo service http
```

**Comment by mrakitin** [comment]
💯 :tada:

**Pull request merged**